File: upload_to_firebase.js
const admin = require('firebase-admin');
const fs = require('fs');

// Initialize Firebase Admin
const serviceAccount = JSON.parse(process.env.FIREBASE_SERVICE_ACCOUNT);

admin.initializeApp({
  credential: admin.credential.cert(serviceAccount)
});

const db = admin.firestore();

async function uploadProducts() {
  console.log('📦 Loading products from JSON...');
  const data = JSON.parse(fs.readFileSync('products_all_brands.json', 'utf8'));
  const products = data.products;

  console.log(`✅ Loaded ${products.length} products`);
  console.log('🚀 Starting upload to Firebase Firestore...');

  const batchSize = 500; // Firestore batch limit
  let uploaded = 0;

  for (let i = 0; i < products.length; i += batchSize) {
    const batch = db.batch();
    const chunk = products.slice(i, i + batchSize);

    chunk.forEach((product) => {
      const docRef = db.collection('products').doc();
      batch.set(docRef, {
        product_title: product.product_title || '',
        product_price: product.product_price || '',
        product_original_price: product.product_original_price || '',
        product_star_rating: product.product_star_rating || '',
        product_num_ratings: parseInt(product.product_num_ratings) || 0,
        product_url: product.product_url || '',
        product_photo: product.product_photo || '',
        platform: product.platform || ''
      });
    });

    try {
      await batch.commit();
      uploaded += chunk.length;
      console.log(`✅ Uploaded ${uploaded}/${products.length} products (${Math.round(uploaded/products.length*100)}%)`);
    } catch (error) {
      console.error(`❌ Error uploading batch ${i}-${i + batchSize}:`, error);
      throw error;
    }

    // Add small delay to avoid rate limiting
    if (i + batchSize < products.length) {
      await new Promise(resolve => setTimeout(resolve, 1000));
    }
  }

  console.log(`🎉 Successfully uploaded ${uploaded} products to Firebase!`);
}

uploadProducts()
  .then(() => {
    console.log('✅ Upload complete!');
    process.exit(0);
  })
  .catch((error) => {
    console.error('❌ Upload failed:', error);
    process.exit(1);
  });

File: scripts/upload_to_firebase.dart
import 'dart:convert';
import 'dart:io';
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:firebase_core/firebase_core.dart';

/// Script to upload 1240+ products to Firebase Firestore
/// Run with: dart run scripts/upload_to_firebase.dart

Future<void> main() async {
  print('=' * 80);
  print('DORON - Firebase Products Upload Script');
  print('=' * 80);

  // Initialize Firebase
  print('\n📱 Initializing Firebase...');
  await Firebase.initializeApp();

  // Read products JSON
  print('\n📂 Reading products.json...');
  final file = File('scripts/products.json');
  final jsonString = await file.readAsString();
  final List<dynamic> productsJson = json.decode(jsonString);

  print('✅ Loaded ${productsJson.length} products');

  // Get Firestore instance
  final firestore = FirebaseFirestore.instance;
  final giftsCollection = firestore.collection('gifts');

  // Upload products in batches
  print('\n🚀 Starting upload to Firebase...');
  print('=' * 80);

  int uploadCount = 0;
  int batchSize = 50;
  int batchNumber = 1;

  for (int i = 0; i < productsJson.length; i += batchSize) {
    final batchEnd = (i + batchSize < productsJson.length)
        ? i + batchSize
        : productsJson.length;

    final batch = productsJson.sublist(i, batchEnd);

    print('\n📦 Batch $batchNumber: Uploading products ${i + 1} to $batchEnd...');

    // Use batch write for better performance
    final writeBatch = firestore.batch();

    for (final productJson in batch) {
      final productData = {
        'id': productJson['id'],
        'brand': productJson['brand'],
        'title': productJson['title'],
        'imageUrl': productJson['imageUrl'],
        'productUrl': productJson['productUrl'],
        'price': productJson['price'],
        'originalPrice': productJson['originalPrice'],
        'category': productJson['category'],
        'tags': List<String>.from(productJson['tags'] ?? []),
        'gender': productJson['gender'],
        'ageRange': productJson['ageRange'],
        'style': productJson['style'],
        'occasion': productJson['occasion'],
        'budgetRange': productJson['budgetRange'],
        'rating': productJson['rating'],
        'numRatings': productJson['numRatings'],
        'verified': productJson['verified'],
        'createdAt': FieldValue.serverTimestamp(),
      };

      // Use product ID as document ID
      final docRef = giftsCollection.doc(productJson['id']);
      writeBatch.set(docRef, productData);
      uploadCount++;
    }

    // Commit batch
    try {
      await writeBatch.commit();
      print('✅ Batch $batchNumber uploaded successfully!');
    } catch (e) {
      print('❌ Error uploading batch $batchNumber: $e');
    }

    batchNumber++;

    // Add small delay between batches to avoid rate limiting
    if (i + batchSize < productsJson.length) {
      await Future.delayed(Duration(milliseconds: 500));
    }
  }

  print('\n' + '=' * 80);
  print('✅ UPLOAD COMPLETE!');
  print('📊 Total products uploaded: $uploadCount');
  print('=' * 80);
}
